### PR TITLE
Optimized Float<M,E> multiplication

### PR DESCRIPTION
**Pull request by huwpascoe**
Based on https://github.com/citra-emu/citra/issues/2939#issuecomment-330706569

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/2944)

### CODE REVIEW

File: src/video_core/pica_types.h [hunk 1]
```diff
@@ -78,12 +78,11 @@ struct Float {
     }
 
     Float<M, E>& operator*=(const Float<M, E>& flt) {
-        if ((this->value == 0.f && !std::isnan(flt.value)) ||
-            (flt.value == 0.f && !std::isnan(this->value)))
-            // PICA gives 0 instead of NaN when multiplying by inf
-            *this = Zero();
-        else
-            value *= flt.ToFloat32();
+        float result = value * flt.ToFloat32();
```
Comment (lioncash) [inline src/video_core/pica_types.h]: Couldn't this be simplified in terms of `operator*` like so:

```cpp
Float<M, E>& operator*=(const Float<M, E>& flt) {
    value = operator*(flt).value;
    return *this;
}
```

**Comment by huwpascoe** [comment]
Changed the *= operator, added assembly output

**Comment by huwpascoe** [comment]
@wwylele done

File: src/video_core/pica_types.h [hunk 2]
```diff
@@ -58,11 +58,12 @@ struct Float {
...
     Float<M, E> operator*(const Float<M, E>& flt) const {
-        if ((this->value == 0.f && !std::isnan(flt.value)) ||
-            (flt.value == 0.f && !std::isnan(this->value)))
-            // PICA gives 0 instead of NaN when multiplying by inf
-            return Zero();
-        return Float<M, E>::FromFloat32(ToFloat32() * flt.ToFloat32());
+        float result = value * flt.ToFloat32();
+        if (!std::isnan(value) && !std::isnan(flt.ToFloat32()))
+            if (std::isnan(result))
+                // PICA gives 0 instead of NaN when multiplying by inf
```
Comment (Subv) [inline src/video_core/pica_types.h]: I'd say move this comment to L62, before the first if.  That'd make it more readable

**Comment by huwpascoe** [comment]
Moved comment

**Pull request merged**